Scales 1.4.0 (#711)
tidyverse/tidyverse.org: content/blog/scales-1-4-0/index.Rmd

photo:
  url: https://unsplash.com/photos/a-close-up-of-a-person-holding-a-paintbrush-Xrelr7cTYm4
  author: Jennie Razumnaya

# one of: "deep-dive", "learn", "package", "programming", "roundup", or "other"
categories: [package] 
tags: [scales]
---

<!--
TODO:
* [x] Look over / edit the post's title in the yaml
* [x] Edit (or delete) the description; note this appears in the Twitter card
* [x] Pick category and tags (see existing with `hugodown::tidy_show_meta()`)
* [x] Find photo & update yaml metadata
* [x] Create `thumbnail-sq.jpg`; height and width should be equal
* [x] Create `thumbnail-wd.jpg`; width should be >5x height
* [x] `hugodown::use_tidy_thumbnails()`
* [x] Add intro sentence, e.g. the standard tagline for the package
* [x] `usethis::use_tidy_thanks()`
-->

We're stoked to announce the release of [scales]({https://scales.r-lib.org/}) 1.4.0. 
scales is a package that provides much of the scaling logic that is used in ggplot2 to a general framework, along with utility functions for e.g. formatting labels or creating colour palettes.

You can install it from CRAN with:

```{r, eval = FALSE}
install.packages("scales")
```

> This blog post will give an overview of the 1.4.0 release, which has some nifty upgrades for working with colours.

Reviewer: ```suggestion
This blog post will give an overview of the 1.4.0 release, which has some nifty upgrades for working with colours and labels.
```

You can see a full list of changes in the [release notes](https://scales.r-lib.org/news/index.html)

```{r setup}
library(scales)
```

```{r ragg, include=FALSE}
knitr::opts_chunk$set(dev = "ragg_png")
```

## Colour manipulation

The `alpha()` and `muted()` functions have been part of scales for a long time. 
Back in the 1.1.0 release we swapped to [farver](https://farver.data-imaginist.com/) to power these functions.
We felt it was appropriate to use this package for other common colour tasks, and so `col_shift()`, `col_lighter()`, `col_darker()`, `col_saturate()` and `col_mix()` were born.

```{r}
my_colours <- c("red", "green", "blue")

m <- rbind(
  original = my_colours,
  shift    = col_shift(my_colours, 90),
  lighter  = col_lighter(my_colours, 20),
  darker   = col_darker(my_colours, 20),
  duller   = col_saturate(my_colours, -50),
  mixed    = col_mix(my_colours, "orchid")
)

show_col(t(m), ncol = ncol(m))
text(x = ncol(m) + 0.25, y = -(1:nrow(m)) + 0.5, rownames(m), adj = 0)
```

## Palettes

> Also palettes have been reworked this release to reflect more useful properties.

Reviewer: ```suggestion
Palettes have also been reworked in this release to include some useful properties.
```

Palettes now come in one of two classes: 'pal_discrete' or 'pal_continuous'.

```{r}
my_palette <- manual_pal(c("palegreen", "deepskyblue", "magenta"))
class(my_palette)
```

> Having palettes as a class rather than plain functions, allows us to store useful metadata about the palette.

Reviewer: ```suggestion
Having palettes as a class rather than as plain functions, allows us to store useful metadata about the palette which can be used downstream.
```

In addition, most colour palette functions also allow the aforementioned colour manipulation functions to work on the palette output.

```{r}
palette_type(my_palette)

palette_nlevels(my_palette)

col_shift(my_palette, 180)(3)
```

> This metadata can then be used to expand discrete palettes to continuous palettes with `as_continuous_pal()` or vise versa to chop up a continuous palette into discrete palettes with `as_discrete_pal()`.

Reviewer: ```suggestion
With the new setup it is now possible to expand discrete palettes to continuous palettes with `as_continuous_pal()` or vise versa to chop up continuous palettes into discrete palettes with `as_discrete_pal()`.
```

```{r}
plot(as_continuous_pal(my_palette))
```

> Another thing to make working with palettes easier, is that the 'scales' package now keeps track of named palettes.

Reviewer: ```suggestion
Another life-improvement for palettes, is that the 'scales' package now keeps track of named palettes.
```

Author: I used 'quality of life improvement' here

By default, the collection of 'known' palettes is pre-populated with colour palettes from the grDevices, RColorBrewer and viridisLite packages.

```{r}
head(palette_names())

get_palette("Okabe-Ito")(8)
```

If you're a developer of a palette package, you can use `set_palette()` to register your palette.

> This has the advantage that your palette is now available to users by name, which at times might be more convenient than having to call the palette generator function.

Reviewer: Maybe something about not having to provide a `scale_*()` function for each palette in the future (once ggplot2 has been updated)

Author: Good call

```{r, error=TRUE}
get_palette("aurora")

set_palette("aurora", my_palette)
plot(get_palette("aurora"))
```

## Labels

> Lastly, please let us introduce you to our two new labelling functions and two new convenience functions for labels.

Reviewer: ```suggestion
This release also provides improvements to labelling in the form of two new labelling functions and two new convenience functions for labels.
```

In contrast to most of scales' label functions, these label functions are great for discrete input.
First up is `label_glue()`, which uses the string interpolation from the glue package to format your labels.

```{r}
label_glue("The {x} penguin")(c("Gentoo", "Chinstrap", "Adelie"))
```

> The next labeling function is convenient when some variable you use consists of shortcodes or abbreviations.

Reviewer: ```suggestion
The other labelling function, `label_dictionary()`, is convenient when some variable you use consists of short-codes or abbreviations.
```

You can provide `label_dictionary()` with a named vector that translates the values to prettier labels.

> If you value didn't exist in the dictionary, these stay as-is by default.

Reviewer: ```suggestion
If one or more of your values doesn't exist in the dictionary, they stay as-is by default.
```

```{r}
dict <- c(
  diy = "Do it yourself", eta = "Estimated time of arrival",
  asap = "As soon as possible", tldr = "Too long; didn't read"
)
label_dictionary(dict)(c("diy", "tldr", "bff"))
```

> The first label convenience function we'd like to tell you about is the `compose_label()` function.

Reviewer: ```suggestion
`compose_label()` is a useful convenience function we've added which will help you to create custom labelling behaviour without needing to write a labelling function from scratch.
```
